uxarray/remap/utils.py

```diff
@@ -0,0 +1,132 @@
+import numpy as np
+
+
+def _remap_grid_parse(
+    source_data, source_grid, destination_grid, coord_type, remap_to, k, query
+):
+    """Gets the destination coordinates from the destination grid for
+    remapping, as well as retrieving the nearest neighbor indices and
+    distances.
+
+    Parameters:
+    -----------
+    source_data : np.ndarray
+        Data variable to remap.
+    source_grid : Grid
+        Source grid that data is mapped from.
+    destination_grid : Grid
+        Destination grid to remap data to.
+    coord_type: str
+        Coordinate type to use for nearest neighbor query, either "spherical" or "Cartesian".
+    remap_to : str
+        Location of where to map data, either "nodes", "edge centers", or "face centers".
+    k : int
+        Number of nearest neighbors to consider in the weighted calculation.
+    query : bool
+        Whether to construct and query the tree based on the source grid.
+
+    Returns:
+    --------
+    dest_coords : np.ndarray
+        Returns the proper destination coordinates based on `remap_to`
+    distances : np.ndarray
+        Returns the distances of the query of `k` nearest neighbors.
+    nearest_neighbor_indices : np.ndarray
+        Returns the nearest neighbor indices of number `k`.
+    """
+
+    n_elements = source_data.shape[-1]
+
+    if n_elements == source_grid.n_node:
+        source_data_mapping = "nodes"
+    elif n_elements == source_grid.n_face:
+        source_data_mapping = "face centers"
+    elif n_elements == source_grid.n_edge:
+        source_data_mapping = "edge centers"
+    else:
+        raise ValueError(
+            f"Invalid source_data shape. The final dimension should match the number of corner "
+            f"nodes ({source_grid.n_node}), edge nodes ({source_grid.n_edge}), or face centers ({source_grid.n_face}) "
+            f"in the source grid, but received: {source_data.shape}"
+        )
+
+    if coord_type == "spherical":
+        if remap_to == "nodes":
+            lon, lat = (
+                destination_grid.node_lon.values,
+                destination_grid.node_lat.values,
+            )
+        elif remap_to == "face centers":
+            lon, lat = (
+                destination_grid.face_lon.values,
+                destination_grid.face_lat.values,
+            )
+        elif remap_to == "edge centers":
+            lon, lat = (
+                destination_grid.edge_lon.values,
+                destination_grid.edge_lat.values,
+            )
+        else:
+            raise ValueError(
+                f"Invalid remap_to. Expected 'nodes', 'edge centers', or 'face centers', "
+                f"but received: {remap_to}"
+            )
+
+        _source_tree = source_grid.get_ball_tree(
+            coordinates=source_data_mapping, reconstruct=True
```

Comment:
Thanks for the review! The `Grid.get_ball_tree` does indeed check and uses the cached version if it's constructed. However, if passed the reconstructed argument, it will override this and ignore the cache. This is needed here because if a user opens a `UxDataset`, and remaps using Cartesian, but then decides to use the same `UxDataset` to remap with spherical, the coordinates used to construct the tree will be incorrect and the query will fail. You can see this in the test cases locally if you remove `reconstruct=True`. 

Also, it could affect the results even if the coordinates aren't changed. The tree is constructed from the grid, so if it's constructed on the nodes on the first remap, and then you remap another variable with the same object that is on the faces, it won't construct the tree on the faces, but use the cached nodes version.

Reply:
This is how we handle caching for our `GeoDataFrame` construction. This may be overkill for what we need here though.

https://github.com/UXARRAY/uxarray/blob/510e29901b898f0554c6054b0367bcb65be99411/uxarray/grid/grid.py#L1567-L1689

Reply:
It does seem like a lot, whenone could just force reconstruct=True.